src/sentry/tasks/deletion/hybrid_cloud.py

```diff
@@ -71,11 +80,19 @@ def set_watermark(
 
 def chunk_watermark_batch(
     prefix: str, field: HybridCloudForeignKey, manager: Manager, *, batch_size: int
-) -> Tuple[int, int, bool, str]:
+) -> WatermarkBatch:
     lower, transaction_id = get_watermark(prefix, field)
     upper = manager.aggregate(Max("id"))["id__max"] or 0
     batch_upper = min(upper, lower + batch_size)
-    return lower, upper, batch_upper + batch_size < upper, transaction_id
+
+    # cap to batch size so that query timeouts don't get us.
+    capped = upper
+    if upper >= batch_upper:
+        capped = batch_upper
+
+    return WatermarkBatch(
+        low=lower, up=capped, has_more=batch_upper < upper, transaction_id=transaction_id
```

Comment:
I've changed the logic for `has_more` to reflect that `upper` is now capped to low + batch size. This should result in multiple queries for small slices of records instead of a single query of 3k elements.